packages/input-phone/src/PhoneInput.hooks.js

```diff
@@ -0,0 +1,118 @@
+import { useRef, useEffect, useMemo, useState, useCallback } from "react";
+import Fuse from "fuse.js";
+
+const options = {
+  shouldSort: true,
+  keys: ["name", "nativeNames", "altSpellings", "flag", "callingCode"],
+};
+
+export const useCountrySearch = (data, query, priority) => {
+  const fuse = useRef(new Fuse(data, options));
+
+  const allCountries = useMemo(() => {
+    try {
+      const codes = priority?.split(",");
+
+      const topCountries = data?.filter(country =>
+        codes?.includes(country.code),
+      );
+
+      const indexedTopCountries = topCountries?.reduce(
+        (acc, country) => ({ ...acc, [country.code]: country }),
+        {},
+      );
+
+      const orderedTopCountries =
+        codes?.map(code => indexedTopCountries[code]) || [];
+
+      return [...orderedTopCountries, ...data];
+    } catch (_error) {
+      // eslint-disable-next-line no-console
+      console.warn(
+        'farmblocks-input-phone: To avoid rerenders, send the priorityCountries list of codes as a string. E.g. "US,CA,GB"',
+      );
+      return data;
+    }
+  }, [data, priority]);
+
+  useEffect(() => {
+    fuse.current = new Fuse(data, options);
+  }, [data]);
+
+  if (!query) {
+    return allCountries;
+  }
+
+  return fuse.current.search(query);
+};
+
+export const useHighlight = ({ items, listRef, selectFn, cancelFn }) => {
+  const [highlightIndex, setHighlightIndex] = useState(-1);
+
+  const changeHighlight = useCallback(
+    modifier => {
+      setHighlightIndex(index => {
+        if (!items.length) {
+          return -1;
+        }
+        const newIndex = index + modifier;
+        if (newIndex < -1) {
+          return items.length - 1;
+        }
+        if (newIndex > items.length - 1) {
+          return -1;
+        }
+        return newIndex;
+      });
+    },
+    [items],
+  );
+
+  useEffect(() => {
+    if (highlightIndex !== -1) {
+      listRef?.current?.scrollToItem?.(highlightIndex);
+    }
+  }, [highlightIndex, listRef]);
+
+  const handleKeyDown = useCallback(
+    event => {
+      const { key } = event;
+      if (key === "Enter") {
+        if (highlightIndex === -1) {
+          return;
+        }
+
+        selectFn(items[highlightIndex].code);
+        return;
+      }
+
+      if (key === "Escape") {
+        setHighlightIndex(-1);
+        cancelFn?.();
```

Comment:
`cancelFn` isn't available on open. We need to query something or browse the list to close the popover on "esc"

Reply:
I closured the `.current` instead of the ref
![](https://media.giphy.com/media/cJBY6cGpla8So1Pv1u/giphy.gif)